### PR TITLE
fix(auth): auth service for all auth and user API endpoints

### PR DESCRIPTION
**Pull request by sbose78**
Note to the reviewer : 
please use the PR title as the commit message after squashing commits during merge.

- Use `auth.openshift.io` for login+user calls instead of `api.openshift.io`
- update login path from `/api/login/authorize` to `/api/login`

This PR has to be merged with-or-after https://github.com/fabric8-ui/ngx-login-client/pull/86 .

### What this PR does ?
- Introduces the usage of auth service endpoints for existing user/token api calls.
- Reads Auth API url from the environment variable.
- The default is https://prod-preview.openshift.io 
- When tested with https://github.com/fabric8-ui/ngx-login-client/pull/86  , the token for github url would no longer be fetched from keycloak but from the auth service.
- When tested with https://github.com/fabric8-ui/ngx-login-client/pull/86  , the refresh token would be fetched from `https://auth.prod-preview.openshift.io/api/token/refresh` instead of  `https://api.prod-preview.openshift.io/api/login/refresh` .
 ### How to test ?
- Checkout https://github.com/fabric8-ui/ngx-login-client/pull/86 , build it and do an `npm link` from the location where the current fabric8-ui branch has been checked out.
- Do an `npm start`
- Verify that all `/api/user` related urls are constructed with the auth service domain. ( except for /api/user/services ) 
- Verify that github token is not fetched from keycloak.

### CODE REVIEW

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-1 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-4 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-5 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by sbose78** [comment]
Hi @joshuawilson  , I would need some help on this - 

When i do an `npm start` locally, this works - 

1. respects the env variable `FABRIC8_AUTH_API_URL` when present 
2. when `FABRIC8_AUTH_API_URL` is absent, uses whatever is defined in https://github.com/fabric8-ui/fabric8-ui/pull/2011/files#diff-66f9bd447d50f2f05180ed487553f32dR28  ( ie, `const FABRIC8_AUTH_API_URL = process.env.FABRIC8_AUTH_API_URL || 'https://auth.prod-preview.openshift.io/api/'`

What more do I need to configure to get urls such as https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io/ working , and also what do I need to do to get this change working in staging + prod.

**Comment by joshuawilson** [comment]
I'll take a look

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-6 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by alexeykazakov** [comment]
Backend is ready to switch UI to auth.openshift.io

**Comment by alexeykazakov** [comment]
New URLs should be used for the following requests:

- **Login:** `api.openshift.io/api/login/authorize` -> `auth.openshift.io/api/login` (path **changed**!)
- **Link accounts:** `api.openshift.io/api/login/linksession` -> `auth.openshift.io/api/link/session` (path **changed**!)
- **Logout:** `api.openshift.io/api/logout` -> `auth.openshift.io/api/logout` 
- **Token refresh:** `api.openshift.io/api/login/refresh` -> `auth.openshift.io/api/token/refresh` (path **changed**!)
- **User search:** `api.openshift.io/api/search/users` -> `auth.openshift.io/api/search/users` 
- **User:** `api.openshift.io/api/user` -> `auth.openshift.io/api/user` 
- **Users:** `api.openshift.io/api/users` -> `auth.openshift.io/api/users` 
- **Collaborators:** `api.openshift.io/api/spaces/{spaceID}/collaborators` -> `auth.openshift.io/api/spaces/{spaceID}/collaborators`

File: environments/openshift-prod-cluster.sh [hunk 1]
```diff
@@ -45,6 +45,8 @@ export WS_K8S_API_SERVER_BASE_PATH=""
 
 export FABRIC8_SSO_API_URL=https://sso.openshift.io/
 export FABRIC8_WIT_API_URL=https://api.openshift.io/api/
+export FABRIC8_AUTH_API_URL=https://api.openshift.io/
```
Comment (alexeykazakov) [inline environments/openshift-prod-cluster.sh]: should be https://**auth**.openshift.io/**api**/, right?

Comment (sbose78) [reply environments/openshift-prod-cluster.sh]: Thanks  - fixed the url too, yet to test out this script driven config.

**Comment by sbose78** [comment]
Thanks for the list @alexeykazakov , 
I will go test them individually and post the results here.

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-7 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-8 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-9 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-11 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by joshuawilson** [comment]
@sbose78 please use semantic release syntax in your commit message

**Comment by sbose78** [comment]
Hi @dlabrecq , Sorry for bothering you in multiple forums :)

When I run this PR with the changes in ngx-login-client PR ( https://github.com/fabric8-ui/ngx-login-client/pull/86 ), the call to `/api/user` fails in the "Getting Started" page. However my changes in ngx-login-client work fine.

I'm not sure what's causing this regression, could you please help ?

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-12 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-13 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by dlabrecq** [comment]
Encountered a problem with the refresh request. The https://auth.prod-preview.openshift.io/api/login/refresh URL does not appear to exist. However, https://api.openshift.io/api/login/refresh does work.

In the comments above, it states that FABRIC8_AUTH_API_URL is respected; otherwise, https://auth.prod-preview.openshift.io/api/ is used. I didn't have FABRIC8_AUTH_API_URL set during this particular test, so this appears to be a bad default URL?

That said, after setting FABRIC8_AUTH_API_URL and clearing my application's local storage, I'm able to get a good github and auth tokens, but the user request is still failing. That's missing the auth header, as described earlier -- still looking at that.

**Comment by sbose78** [comment]
Thanks @dlabrecq 

> Encountered a problem with the refresh request. The https://auth.prod-preview.openshift.io/api/login/refresh URL does not appear to exist. However, https://api.openshift.io/api/login/refresh does work.

This change here should use the new `/token/refresh` url of the Auth service ( https://auth.prod-preview.openshift.io ) irrespective of whether the env var `FABRIC8_AUTH_API_URL` is set.
https://github.com/fabric8-ui/ngx-login-client/pull/86/files#diff-9a5b1e9ba2deaa637efb0e34749061aeR116

**Comment by dlabrecq** [comment]
I did notice that if HttpService of ngx-login-client cannot find auth_token in local storage, the auth header is not included. I suspect this may still have something do do with the bad refresh request?

**Comment by sbose78** [comment]
Thanks @dgutride 

. In the comments above, it states that FABRIC8_AUTH_API_URL is respected; otherwise, https://auth.prod-preview.openshift.io/api/ is used. I didn't have FABRIC8_AUTH_API_URL set during this particular test, so this appears to be a bad default URL?

I tried the above without setting anything up in my new laptop - I can see the default being respected and I can see token refresh happening smoothly with the changes in `ngx-login-client` . Could you retry it later maybe?

> I did notice that if HttpService of ngx-login-client cannot find auth_token in local storage, the auth header is not included. I suspect this may still have something do do with the bad refresh request? I believe if the request is bad, some properties in local storage may be cleared?

I checked the other requests originating from the browser and strangely all of them have the auth header :(  even the refresh token works `/token/refresh`. Are you able to see requests returning a *200OK* to `https://auth.prod-preview.openshift.io/api/token/refresh` after you set the env var ?

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-14 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by dlabrecq** [comment]
I tested your latest commit and the UI appears to be working better. I'm no longer seeing a bad user request.

I still see a bad refresh OPTIONS request, tho. That request is using https://auth.openshift.io/api/login/refresh. Tested that in Postman and it doesn't appear to be a valid URL. 

Note that https://api.openshift.io/api/login/refresh works.

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-15 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

File: config/webpack.prod.js [hunk 2]
```diff
@@ -36,7 +36,7 @@ const FABRIC8_WIT_API_URL = process.env.FABRIC8_WIT_API_URL;
 const FABRIC8_REALM = process.env.FABRIC8_REALM;
 const FABRIC8_RECOMMENDER_API_URL = process.env.FABRIC8_RECOMMENDER_API_URL || 'http://api-bayesian.dev.rdu2c.fabric8.io/api/v1/';
 const FABRIC8_SSO_API_URL = process.env.FABRIC8_SSO_API_URL;
-const FABRIC8_AUTH_API_URL = process.env.FABRIC8_AUTH_API_URL;
+const FABRIC8_AUTH_API_URL = process.env.FABRIC8_AUTH_API_URL || 'http://auth.prod-preview.openshift.io/api/';
```
Comment (alexeykazakov) [inline config/webpack.prod.js]: use `https://`

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-16 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

File: config/webpack.prod.js (hunk 2 above)
Comment (sbose78) [reply config/webpack.prod.js]: thanks , it was a temporary commit - removed.

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-17 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-18 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@sbose78 fabric8/fabric8-ui:SNAPSHOT-PR-2011-19 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2011-fabric8-ui.badger.fabric8.io

**Comment by sbose78** [comment]
Hi @dlabrecq , this PR is ready for review.

**Comment by joshuawilson** [comment]
@sbose78 @alexeykazakov I have a PR that might be what we need for env var.

https://github.com/fabric8io/fabric8-pipeline-library/pull/356

**Pull request merged**